core/src/main/java/org/ldaptive/dn/RDn.java

```diff
@@ -0,0 +1,212 @@
+   *
+   * @param  rdn  to parse
+   *
+   * @throws  IllegalArgumentException  if rdn contains multiple RDNs or no RDNs
+   */
+  public RDn(final String rdn)
+  {
+    this(rdn, new DefaultDnParser());
+  }
+
+
+  /**
+   * Creates a new RDN with the supplied string.
+   *
+   * @param  rdn  to parse
+   * @param  parser  to parse dn
+   *
+   * @throws  IllegalArgumentException  if rdn contains multiple RDNs or no RDNS
+   */
+  public RDn(final String rdn, final DnParser parser)
+  {
+    final List<RDn> rdns = parser.parse(rdn);
+    if (rdns.isEmpty()) {
+      throw new IllegalArgumentException("Invalid RDN: no RDNs found in " + rdn);
+    }
+    if (rdns.size() > 1) {
+      throw new IllegalArgumentException("Invalid RDN: multiple RDNs found in " + rdn);
+    }
+    nameValues = rdns.get(0).getNameValues();
+  }
+
+
+  /**
+   * Creates a new RDN with the supplied name value pairs.
+   *
+   * @param  value  to add
+   */
+  public RDn(final NameValue... value)
+  {
+    nameValues = Stream.of(value).filter(Objects::nonNull).collect(
+      Collectors.collectingAndThen(Collectors.toCollection(LinkedHashSet::new), Collections::unmodifiableSet));
+  }
+
+
+  /**
+   * Creates a new RDN with the supplied name value pairs.
+   *
+   * @param  values  to add
+   */
+  public RDn(final Collection<NameValue> values)
+  {
+    nameValues = values.stream().filter(Objects::nonNull).collect(
+      Collectors.collectingAndThen(Collectors.toCollection(LinkedHashSet::new), Collections::unmodifiableSet));
+  }
+
+
+  /**
+   * Creates a new RDN with a single name value pair.
+   *
+   * @param  attributeName  to add
+   * @param  attributeValue  to add
+   */
+  public RDn(final String attributeName, final String attributeValue)
+  {
+    nameValues = Set.of(new NameValue(attributeName, attributeValue));
+  }
+
+
+  /**
+   * Returns the first name value pair in this RDN.
+   *
+   * @return  name value pair
+   */
+  public NameValue getNameValue()
+  {
+    if (nameValues.isEmpty()) {
+      return null;
+    }
+    return nameValues.iterator().next();
+  }
+
+
+  /**
+   * Returns all the name value pairs in this RDN.
+   *
+   * @return  name value paris
+   */
+  public Set<NameValue> getNameValues()
+  {
+    return nameValues;
+  }
+
+
+  /**
+   * Returns the number of name value pairs in this RDN.
+   *
+   * @return  RDN size
+   */
+  public int size()
+  {
+    return nameValues.size();
+  }
+
+
+  /**
+   * Returns a string representation of this RDN, joining each name value pair with '+'.
+   *
+   * @return  string form of the RDN
+   */
+  public String format()
+  {
+    return format(null);
+  }
+
+
+  /**
+   * Returns a string representation of this RDN, joining each name value pair with '+'.
+   *
+   * @param  normalizer  to apply to the RDN components or null for no formatting
+   *
+   * @return  string form of the RDN
+   */
+  public String format(final RDnNormalizer normalizer)
```

Comment:
Perhaps formatting should also be pushed out into a dedicated component for symmetry with parsing. This is the approach of may libraries I've seen, which doesn't preclude convenience methods like this, but it's a good way of separating concerns without compromising convenience. `java.time` date parsing/formatting comes to mind as an example of this pattern.

Reply:
I considered that, but there's so little for the formatter to do. The complexity is in the normalizer, which is already externalized.